nixos/modules/security/acme/default.nix

```diff
@@ -15,34 +14,45 @@ let
   mkAccountHash = acmeServer: data: mkHash "${toString acmeServer} ${data.keyType} ${data.email}";
   accountDirRoot = "/var/lib/acme/.lego/accounts/";
 
-  lockdir = "/run/acme/";
-  concurrencyLockfiles = map (n: "${toString n}.lock") (lib.range 1 cfg.maxConcurrentRenewals);
-  # Assign elements of `baseList` to each element of `needAssignmentList`, until the latter is exhausted.
-  # returns: [{fst = "element of baseList"; snd = "element of needAssignmentList"}]
-  roundRobinAssign = baseList: needAssignmentList:
-    if baseList == [] then []
-    else _rrCycler baseList baseList needAssignmentList;
-  _rrCycler = with builtins; origBaseList: workingBaseList: needAssignmentList:
-    if (workingBaseList == [] || needAssignmentList == [])
-    then []
-    else
-      [{ fst = head workingBaseList; snd = head needAssignmentList;}] ++
-      _rrCycler origBaseList (if (tail workingBaseList == []) then origBaseList else tail workingBaseList) (tail needAssignmentList);
-  attrsToList = lib.mapAttrsToList (attrname: attrval: {name = attrname; value = attrval;});
-  # for an AttrSet `funcsAttrs` having functions as values, apply single arguments from
-  # `argsList` to them in a round-robin manner.
-  # Returns an attribute set with the applied functions as values.
-  roundRobinApplyAttrs = funcsAttrs: argsList: lib.listToAttrs (map (x: {inherit (x.snd) name; value = x.snd.value x.fst;}) (roundRobinAssign argsList (attrsToList funcsAttrs)));
-  wrapInFlock = lockfilePath: script:
-    # explainer: https://stackoverflow.com/a/60896531
-    ''
-      exec {LOCKFD}> ${lockfilePath}
-      echo "Waiting to acquire lock ${lockfilePath}"
-      ${pkgs.flock}/bin/flock ''${LOCKFD} || exit 1
-      echo "Acquired lock ${lockfilePath}"
-    ''
-    + script + "\n"
-    + ''echo "Releasing lock ${lockfilePath}"  # only released after process exit'';
+  writeMaxConcurrentShellScript = name: text:
+    let script = pkgs.writeShellScript name text; in
+    if cfg.maxConcurrentRenewals == 0
+    then script
+    else "exec ${lib.escapeShellArg lockAndRun} ${lib.escapeShellArg script}";
+
+  lockdir = "/run/acme";
+  lockAndRun = pkgs.writeShellScript "acme-lock-and-run.sh" ''
+    ${lib.optionalString cfg.defaults.enableDebugLogs "set -x"}
+    set -euo pipefail
+
+    echo >&2 'Waiting to acquire lock'
+
+    # Successively try each lock until we acquire one
+    i="$RANDOM"  # Reduce contention/average wait
+    while true; do
+      lock=${lib.escapeShellArg lockdir}/"$((i = (i + 1) % ${toString cfg.maxConcurrentRenewals})).lock"
```

Comment:
Just for the sake of readability, could you perform the variable assignment on its own line?

Reply:
Just to add - I really like the use of $RANDOM and the fact that it will try all possible locks. This is way more efficient than before, definitely more performant than what I previously proposed with systemd dependency chaining.